sncosmo/models.py

```diff
@@ -2030,3 +2030,95 @@ def propagate(self, wave, flux, phase=None):
         """Propagate the flux."""
         ebv = self._parameters[0]
         return extinction.apply(self._f(wave, ebv * self._r_v), flux)
+
+
+class G10(PropagationEffect):
+    """Guy (2010) SNe Ia non-coherent scattering.
+    
+    Implementation is done following arxiv:1209.2482."""
+
+    _param_names = ['L0', 'F0', 'F1', 'dL']
+    param_names_latex = [r'\lambda_0', 'F_0', 'F_1', 'd_L']
+
+    def __init__(self, SALTsource):
+        """Initialize G10 class."""
+        self._parameters = np.array([2157.3, 0.0, 1.08e-4, 800])
+        self._colordisp = SALTsource._colordisp
+        self._minwave = SALTsource.minwave()
+        self._maxwave = SALTsource.maxwave()
+
+    def compute_sigma_nodes(self):
+        """Computes the sigma nodes."""
+        L0, F0, F1, dL = self._parameters
+        lam_nodes = np.arange(self._minwave, self._maxwave, dL)
+        if lam_nodes.max() < self._maxwave:
+            lam_nodes = np.append(lam_nodes, self._maxwave)
+        siglam_values = self._colordisp(lam_nodes) 
+        
+        siglam_values[lam_nodes < L0] *= 1 + (lam_nodes[lam_nodes < L0] - L0) * F0
+        siglam_values[lam_nodes > L0] *= 1 + (lam_nodes[lam_nodes > L0] - L0) * F1
+        siglam_values *= np.random.normal(size=len(lam_nodes))
+        
+        return lam_nodes, siglam_values
+
+    def propagate(self, wave, flux):
+        """Propagate the effect to the flux."""
+        lam_nodes, siglam_values = self.compute_sigma_nodes()
```

Comment:
Would moving this to `__init__` make this `SALTwithG10.bandflux` the same on repeated calls?

Reply:
Yes it should works since the random part is in `self.compute_sigma_nodes()`

Reply:
Moving this part in `__init__ `  leads to a non-functional `Model.set()` since the updates on `self._parameters` will not be taken into account. I set a `self._seed` parameter in the `__init__` to fix the randomness when init the model.